File: app/src/main/java/com/dumuzeyn/mp3player/LibraryFileAccessManager.java
package com.dumuzeyn.mp3player;

import android.content.Context;
import com.dumuzeyn.mp3player.ui.permissions.DeviceAudioPermissionController;

final class LibraryFileAccessManager {
    enum AccessState {
        AVAILABLE,
        UNAVAILABLE,
        DEFERRED
    }

    private LibraryFileAccessManager() {
    }

    static AccessState accessState(Context context, Track track) {
        if (track == null || track.uri == null || track.uri.trim().isEmpty()) {
            return AccessState.UNAVAILABLE;
        }
        android.net.Uri uri;
        try {
            uri = track.asUri();
        } catch (RuntimeException invalidUri) {
            return AccessState.UNAVAILABLE;
        }
        String authority = uri.getAuthority();
        if ("media".equalsIgnoreCase(authority)
                && !DeviceAudioPermissionController.hasPermission(context)) {
            return AccessState.DEFERRED;
        }
        return TrackStore.canOpenForRead(context, uri)
                ? AccessState.AVAILABLE : AccessState.UNAVAILABLE;
    }

}

File: app/src/main/java/com/dumuzeyn/mp3player/LibraryFileAccessManager.kt
package com.dumuzeyn.mp3player

import android.content.Context
import com.dumuzeyn.mp3player.ui.permissions.DeviceAudioPermissionController

object LibraryFileAccessManager {
    enum class AccessState { AVAILABLE, UNAVAILABLE, DEFERRED }

    @JvmStatic
    fun accessState(context: Context, track: Track?): AccessState {
        if (track == null || track.uri.isBlank()) return AccessState.UNAVAILABLE
        val uri = runCatching(track::asUri).getOrElse { return AccessState.UNAVAILABLE }
        if (
            uri.authority.equals("media", ignoreCase = true) &&
            !DeviceAudioPermissionController.hasPermission(context)
        ) {
            return AccessState.DEFERRED
        }
        return if (TrackStore.canOpenForRead(context, uri)) {
            AccessState.AVAILABLE
        } else {
            AccessState.UNAVAILABLE
        }
    }
}
